providers/providers.go

```diff
@@ -185,69 +191,49 @@ func writeProviderEntry(dstore ds.Datastore, k cid.Cid, p peer.ID, t time.Time)
 	return dstore.Put(ds.NewKey(dsk), buf[:n])
 }
 
-func (pm *ProviderManager) deleteProvSet(k cid.Cid) error {
-	pm.providers.Remove(k.KeyString())
-
+func (pm *ProviderManager) gc() {
 	res, err := pm.dstore.Query(dsq.Query{
 		KeysOnly: true,
-		Prefix:   mkProvKey(k),
+		Prefix:   providersKeyPrefix,
 	})
 	if err != nil {
-		return err
-	}
-
-	entries, err := res.Rest()
-	if err != nil {
-		return err
+		log.Error("error garbage collecting provider records: ", err)
+		return
 	}
+	defer res.Close()
 
-	for _, e := range entries {
-		err := pm.dstore.Delete(ds.NewKey(e.Key))
-		if err != nil {
-			log.Error("deleting provider set: ", err)
+	now := time.Now()
+	for {
+		e, ok := res.NextSync()
+		if !ok {
+			return
 		}
-	}
-	return nil
-}
-
-func (pm *ProviderManager) getProvKeys() (func() (cid.Cid, bool), error) {
-	res, err := pm.dstore.Query(dsq.Query{
-		KeysOnly: true,
-		Prefix:   providersKeyPrefix,
-	})
-	if err != nil {
-		return nil, err
-	}
 
-	iter := func() (cid.Cid, bool) {
-		for e := range res.Next() {
-			parts := strings.Split(e.Key, "/")
-			if len(parts) != 4 {
-				log.Warningf("incorrectly formatted provider entry in datastore: %s", e.Key)
-				continue
-			}
-			decoded, err := base32.RawStdEncoding.DecodeString(parts[2])
-			if err != nil {
-				log.Warning("error decoding base32 provider key: %s: %s", parts[2], err)
-				continue
-			}
+		if e.Error != nil {
+			log.Error("got an error: ", e.Error)
+			continue
+		}
 
-			c, err := cid.Cast(decoded)
+		// check expiration time
+		t, err := readTimeValue(e.Value)
+		switch {
+		case err != nil:
+			// couldn't parse the time
+			log.Warning("parsing providers record from disk: ", err)
+			fallthrough
+		case now.Sub(t) > ProvideValidity:
+			// or just expired
+			err = pm.dstore.Delete(ds.RawKey(e.Key))
```

Comment:
shouldn't we also clean the lru cache here? Alternatively we can simply flush it completely on gc.

Reply:
Ah, there is a `Purge` call before entering gc, good.